### PR TITLE
Contact setting and rough epp management

### LINKED ISSUE #562: Connect up nameserver management page with EPP commands
Ensure NameServers are set on save.

NOTE: Testing interactions of our Django app and the results of EPP commands is not something we have figured out before, probably some extra work there.

This should be completed after #493 .
relates to #848

### LINKED ISSUE #808: Delete a domain after it has transitioned through clientHold
### Story

As an analyst
I need to delete a domain that has gone through the clientHold transition state.
so that it's removed from the .gov zone.

### Acceptance Criteria

- [ ] add a delete button (EDIT: update the existing delete domain button as needed)
- [ ] deleting can only be done if the domain is in the state HOLD and should be done using the deleted() function (currently already added to the button

_Note:_
_- the Domain application should still show_
_- the this should not delete the domain object, only the status_

### Additional Context

The delete request from a registrant will be sent via email to CISA analysts and is not in scope of this feature. The analyst will be responsible for clicking client hold first before clicking the delete button. Do not need to include a warning pop-up in this ticket, that can be added later if desired at all. The same goes for indicating to them that client hold is needed first. 

note: this can be broken up if epp centric logic is done separately. When implementing the button portion, the developer  can copy logic existing for the place hold button on the django-admin domain object.

Refinement Question: 
Suggested approaches for what happens to the application afterwards:
1. the application will still show as "approved" but that the domain management section is not viewable and the manage icon is removed. A column will be included that contains the status of the domain object
Needs a new ticket @kitsushadow 

### Issue Links

Relates to #803 
blocked by: #901

### LINKED ISSUE #964: Update "Domain State" in Registrar to reflect the State Diagram.
As a developer,
I want to have the possible states for a Domain Request or Domain object exactly match those states described in the referenced States diagram,
So that I may use these states in my code.

We have defined and updated Domain Request and Domain states in the authoritative document here:
[Miro - States Visual](https://miro.com/app/board/uXjVMuqbLOk=/?share_link_id=56664338997)

Please update the Registrar to reflect the possible states noted in the above document.

AC:

- [ ] All states in the Registrar system, as described in the state diagram, are defined and available for use.
- [ ] No states are defined that are not referenced in the state diagram (to avoid confusion)

Relates to #803
Relates to #929

### PR DESCRIPTION
**Pull request by abroddrick**
## Resolves #808 
Resolves #562 (see note)
Resolves #964 
Implements the bare bones of:
#901 
#857 
#851 
#850 
#848 
#847 
#437 
adds todos and some error handling for #433 
some for #912 

## Changes

- contact getting implemented
- updates to the cache function
- basic delete, client hold, remove clienthold and related transitions implemented
- all domain states  updated to match
- retrieves the domain statuses

## Context for reviewers
TLDR: This is meant to push up the contact setting piece and get this code more viewable. That is why not every piece is implemented fully. The one that should be considered for reviewing mostly is the contact getting piece and its related tests (utlilizing security contact). Assume everything else is minimally tested. 

for 562 should I disconnect the connection, probably is buggy at the moment 

-When going to the domain management page, the check already there for domain.nameservers now results in the domain being created in EPP and the state switching to DNS NEEDED (if not already)
-Nameserver page will now show no nameservers for domains that haven't had them added
-going to the domain nameserver page you can start to add nameservers. NOTE: this is not tested a ton as that is #848. Let me know your thoughts on keeping this partially implemented for the short while until 848 can be finished
-When the domain is created all contacts are created, including security
-you will not see the default security contact on the domain management page as that still uses just DomainInformation

## Setup

OPTIONAL:
for developers run the following commands to test manually in a my sandbox:
Note: My sandbox isn't loading fixutres so you can't access admin. Feel free to change mydomain to a unique domain if desired. Also the hosts requires a check on if the host is viewed as a available on the domain, so if one is taken it is not added, just try new ones if that happens. I have also had weird stuff happen when I type None in shell in which case use the tuple notation ("nameserver.com", ) instead of using None.

cf push getgov-ab -f ops/manifests/manifest-ab.yaml
 cf ssh getgov-ab 
 /tmp/lifecycle/shell
from epplib.models import common
from epplibwrapper import CLIENT as registry, commands
from registrar.models.domain import Domain

domain, _ = Domain.objects.get_or_create(name="myDomain.gov")
domain._get_or_create_domain()

hosts=[("ns5.example.com"”,None),(”ns6.example.com","0.0.0.0") ]
domain.nameservers=hosts

_Now print the data to see what happened_
status on the domain
print(domain.statuses)

nameservers on the domain
print(domain.nameservers)

_get contacts_
request = commands.InfoDomain(name="myDomain.gov")
response= registry.send(request, cleaned=True)

print(response.res_date[0].contacts)
print(response.res_data[0].registrant)

Otherwise:

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Comment by zandercymatics** [comment]
@abroddrick  I don't believe I currently have space-level access to your sandbox (not an applicable space - cannot run commands on it or see logs), though I can access it through the url fine (though with the /admin issue you mentioned), I'll take a look through your code for now and pull this locally

File: src/epplibwrapper/client.py [hunk 1]
```diff
@@ -83,7 +83,7 @@ def _send(self, command):
             logger.warning(message, cmd_type, exc_info=True)
             raise RegistryError(message) from err
         except Exception as err:
-            message = "%s failed to execute due to an unknown error."
+            message = "%s failed to execute due to an unknown error." % err
```
Comment (zandercymatics) [inline src/epplibwrapper/client.py:86]: Nice catch

File: src/registrar/admin.py [hunk 2]
```diff
@@ -192,6 +205,139 @@ def response_change(self, request, obj):
...
             return HttpResponseRedirect(".")
+
+        if GET_SECURITY_EMAIL in request.POST:
+            try:
+                contacts = obj._get_property("contacts")
+                email = None
+                for contact in contacts:
+                    if ["type", "email"] in contact.keys() and contact[
+                        "type"
+                    ] == "security":
```
Comment (zandercymatics) [inline src/registrar/admin.py:216]: Weird linter thing?

Comment (abroddrick) [reply src/registrar/admin.py:216]: @zandercymatics please don't review this yet. I am still making changes that is why this is set to draft. And yes that is for sure black formatting

File: src/registrar/admin.py [hunk 3]
```diff
@@ -192,6 +203,138 @@ def response_change(self, request, obj):
                     % obj.name,
                 )
             return HttpResponseRedirect(".")
+
+        if GET_SECURITY_EMAIL in request.POST:
```
Comment (zandercymatics) [inline src/registrar/admin.py:207]: Untangling this will be a bit of a headache, is this something you'd want me to help you on? I've dealt with this particular function enough to understand how it works pretty well so just separating all of this out shouldn't take me too long

File: src/registrar/management/commands/load.py [hunk 4]
```diff
@@ -13,11 +13,8 @@ class Command(BaseCommand):
     def handle(self, *args, **options):
         # django-auditlog has some bugs with fixtures
         # https://github.com/jazzband/django-auditlog/issues/17
-        if settings.DEBUG:
```
Comment (zandercymatics) [inline src/registrar/management/commands/load.py:16]: I do actually recall this getting added with the fixtures change... This makes sense. This raises the question of how we should handle fixture data for production environments, since we probably don't want to be loading this in prod. Regardless, this is outside this PR imo

File: src/registrar/models/domain.py [hunk 5]
```diff
@@ -96,7 +96,7 @@ class Status(models.TextChoices):
         # for human review or third-party action.  A transform command that
         # is processed, but whose requested action is pending, is noted with
         # response code 1001.
-        PENDING_CREATE = "pendingCreate"
+        DNS_NEEDED = "pendingCreate"
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:99]: Why was this changed?

File: src/registrar/models/domain.py [hunk 6]
```diff
@@ -219,17 +226,107 @@ def nameservers(self) -> list[tuple[str]]:
...
         """
-        # TODO: call EPP to get this info instead of returning fake data.
-        return [
-            ("ns1.example.com",),
-            ("ns2.example.com",),
-            ("ns3.example.com",),
-        ]
+        try:
+            hosts = self._get_property("hosts")
+        except Exception as err:
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:231]: In future after we implement this, we will likely want to change this so that it catches a specific registry error rather than a broad stroke one (as we'd want to log that seperately)

**Review by zandercymatics — COMMENTED** [review]
Will add more comments later...

File: src/registrar/models/domain.py [hunk 7]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+    def _update_epp_contact(self, contact: PublicContact):
+        """Sends UpdateContact to update the actual contact object,
+        domain object remains unaffected
+        should be used when changing email address
+        or other contact info on an existing domain
+        """
+        updateContact = commands.UpdateContact(
+            id=contact.registry_id,
+            # type: ignore
+            postal_info=self._make_epp_contact_postal_info(contact=contact),
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:399]: Nice helper. May use this as part of the getter ticket

File: src/registrar/models/domain.py [hunk 8]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+        logger.info(
+            "_update_domain_with_contact() received type %s " % contact.contact_type
+        )
+        domainContact = epp.DomainContact(
+            contact=contact.registry_id, type=contact.contact_type
+        )
+
+        updateDomain = commands.UpdateDomain(name=self.name, add=[domainContact])
+        if rem:
+            updateDomain = commands.UpdateDomain(name=self.name, rem=[domainContact])
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:424]:
```suggestion
        # May be invalid syntax but I think you can do this...
        updateDomain: commands.UpdateDomain
        if rem:
            updateDomain = commands.UpdateDomain(name=self.name, rem=[domainContact])
        else:
            updateDomain = commands.UpdateDomain(name=self.name, add=[domainContact])
```
If rem=True, wouldn't this add the contact then immediately delete it? Seems potentially inefficient if not intentional

File: src/registrar/models/domain.py [hunk 9]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+        """Get or set the security contact for this domain."""
+
+        # get the contacts: call _get_property(contacts=True)
+        # if contacts exist and security contact is in the contact list
+        # return that contact
+        # else call the setter
+        #   send the public default contact
+        try:
+            contacts = self._get_property("contacts")
+            for contact in contacts:
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:452]: Should we be accounting for an array of security_contacts to be returned here? I was under the assumption that, while possible, for our purposes we will only deal with one security_contact (and others in parallel) per domain

File: src/registrar/models/domain.py [hunk 10]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+                ##zander don't do this just to do the bare bones here
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:453]: I'd assume by barebones you mean: contacts = self._get_property("contacts")?

File: src/registrar/models/domain.py [hunk 11]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+        try:
+            registry.send(updateDomain, cleaned=True)
+        except RegistryError as e:
+            logger.error(
+                "Error changing to new registrant error code is %s, error is %s"
+                % (e.code, e)
+            )
+            # TODO-error handling better here?
+
+    def _set_singleton_contact(self, contact: PublicContact, expectedType: str):  # noqa
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:482]: In the function description, I'd recommend specifying what expectedType is used for in this context (an example would be useful here)

File: src/registrar/models/domain.py [hunk 12]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+        )
+
+        # get publicContact objects that have the matching
+        # domain and type but a different id
+        # like in highlander we there can only be one
+        hasOtherContact = (
+            PublicContact.objects.exclude(registry_id=contact.registry_id)
+            .filter(domain=self, contact_type=contact.contact_type)
+            .exists()
+        )
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:509]: This is being used again in existing contact:
```
            if isEmptySecurity:
                existing_contact = PublicContact.objects.filter(
                    domain=self, contact_type=contact.contact_type
                ).get()
            else:
                existing_contact = (
                    PublicContact.objects.exclude(registry_id=contact.registry_id)
                    .filter(domain=self, contact_type=contact.contact_type)
                    .get()
                )
```
I believe that (and I may be wrong), you can call `.exclude` _after_ `.filter`. If so, then we are doing this filter operation twice unnecessarily. You can create an object which stores the following:
`filteredContacts = PublicContact.objects.filter(domain=self, contact_type=contact.contact_type)` 
This allows you to refactor this check to this (ignore the weird github formatting):
 ```suggestion
        filteredContacts = PublicContact.objects.filter(domain=self, contact_type=contact.contact_type)
        hasOtherContact = (
            filteredContacts.exclude(registry_id=contact.registry_id).exists()
        )
```

More on this below....

File: src/registrar/models/domain.py [hunk 13]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+                existing_contact = PublicContact.objects.filter(
+                    domain=self, contact_type=contact.contact_type
+                ).get()
+            else:
+                existing_contact = (
+                    PublicContact.objects.exclude(registry_id=contact.registry_id)
+                    .filter(domain=self, contact_type=contact.contact_type)
+                    .get()
+                )
+            if isRegistrant:
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:543]: By storing filtered contacts in a variable as described above (`filteredContacts = PublicContact.objects.filter(domain=self, contact_type=contact.contact_type`), this also allows you to refactor these as so (ignore the weird github formatting... may need to correct it):
```suggestion
            if isEmptySecurity:
                   filteredContacts.get()
            else:
                existing_contact = (
                    filteredContacts.exclude(registry_id=contact.registry_id).get()
                )
            if isRegistrant:
```
You could take this further if you want and store the .exclude operation in a variable, but this should prevent .filter being called twice, as we should only need to do that once here

**Review by zandercymatics — COMMENTED** [review]
Here are some additional suggestions on this

File: src/registrar/models/domain.py [hunk 14]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+
+        # contact doesn't exist on the domain yet
+        logger.info("_set_singleton_contact()-> contact has been added to the registry")
+
+        # if has conflicting contacts in our db remove them
+        if hasOtherContact:
+            logger.info(
+                "_set_singleton_contact()-> updating domain, removing old contact"
+            )
+            if isEmptySecurity:
```
Comment (abroddrick) [inline src/registrar/models/domain.py:533]: past me wrong

File: src/registrar/models/domain.py [hunk 15]
```diff
@@ -219,17 +226,107 @@ def nameservers(self) -> list[tuple[str]]:
...
+                # update the domain obj
+                if createdCode == ErrorCode.COMMAND_COMPLETED_SUCCESSFULLY:
+                    # add host to domain
+                    request = commands.UpdateDomain(
+                        name=self.name, add=[epp.HostObjSet([host])]
+                    )
+
+                    try:
+                        registry.send(request, cleaned=True)
+                    except RegistryError as e:
```
Comment (abroddrick) [inline src/registrar/models/domain.py:314]: needs to be handled better in nameserver ticket, or in 433

File: src/registrar/models/domain.py [hunk 16]
```diff
@@ -381,78 +719,226 @@ def _validate_host_tuples(self, hosts: list[tuple[str]]):
...
+        logger.info("pendingCreate()-> inside pending create")
+        self._delete_domain()
+        # TODO - delete ticket any additional error handling here
+
+    @transition(
+        field="state",
+        source=[State.DNS_NEEDED],
+        target=State.READY,
+    )
+    def ready(self):
```
Comment (abroddrick) [inline src/registrar/models/domain.py:823]: consider adding the contents of the if as a condition for transitioning. check how this effects error handling and messaging for the user.

File: src/registrar/admin.py (hunk 3 above)
Comment (abroddrick) [reply src/registrar/admin.py:207]: delete for me please

File: src/registrar/admin.py [hunk 17]
```diff
@@ -192,6 +203,138 @@ def response_change(self, request, obj):
...
+            else:
+                self.message_user(
+                    request,
+                    ("Domain %s is now in clientHold") % obj.name,
+                )
+            return HttpResponseRedirect(".")
+
+        elif REMOVE_CLIENT_HOLD in request.POST:
+            try:
+                obj.revertClientHold()
```
Comment (abroddrick) [inline src/registrar/admin.py:315]: remove my implementation and use Daves for remove and set client hold

File: src/registrar/admin.py [hunk 18]
```diff
@@ -192,6 +203,138 @@ def response_change(self, request, obj):
...
+            except Exception as err:
+                self.message_user(request, err, messages.ERROR)
+            else:
+                self.message_user(
+                    request,
+                    ("Nameservers are %s" ". Thanks!") % nameservers,
+                )
+            return HttpResponseRedirect(".")
+
+        elif GET_STATUS in request.POST:
```
Comment (abroddrick) [inline src/registrar/admin.py:288]: keep please

File: src/registrar/admin.py [hunk 19]
```diff
@@ -192,6 +203,138 @@ def response_change(self, request, obj):
...
+            except Exception as err:
+                self.message_user(request, err, messages.ERROR)
+            else:
+                self.message_user(
+                    request,
+                    ("Domain created with %s" ". Thanks!") % obj.name,
+                )
+            return HttpResponseRedirect(".")
+
+        elif MAKE_NAMESERVERS in request.POST:
```
Comment (abroddrick) [inline src/registrar/admin.py:264]: delete this

File: src/registrar/admin.py [hunk 20]
```diff
@@ -192,6 +203,138 @@ def response_change(self, request, obj):
...
+            except Exception as err:
+                self.message_user(request, err, messages.ERROR)
+            else:
+                self.message_user(
+                    request,
+                    ("Domain %s will now have client hold removed") % obj.name,
+                )
+            return HttpResponseRedirect(".")
+
+        elif DELETE_DOMAIN in request.POST:
```
Comment (abroddrick) [inline src/registrar/admin.py:326]: keep

File: src/registrar/models/domain.py [hunk 21]
```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
...
+            if isRegistrant:
+                # send update domain only for registant contacts
+                existing_contact.delete()
+                self._add_registrant_to_existing_domain(contact)
+            else:
+                # remove the old contact and add a new one
+                try:
+                    self._update_domain_with_contact(contact=existing_contact, rem=True)
+                    existing_contact.delete()
+                    if isEmptySecurity:
```
Comment (abroddrick) [inline src/registrar/models/domain.py:552]: delete

File: src/registrar/models/domain.py (hunk 6 above)
Comment (abroddrick) [reply src/registrar/models/domain.py:234]: It will throw a value error here, or a registry error so there will be a bit of error handling needed for those cases. But agreed that is  covered in the nameservers ticket and 433 which handles error handling

File: src/registrar/models/domain.py (hunk 8 above)
Comment (abroddrick) [reply src/registrar/models/domain.py:427]: No the command isn't sent yet, it is just making a command and avoiding an else block basically

File: src/registrar/models/domain.py (hunk 9 above)
Comment (abroddrick) [reply src/registrar/models/domain.py:455]: This is an array of all the contacts, it hasn't retrieved the security one yet. Good question though!

File: src/registrar/models/domain.py (hunk 10 above)
Comment (abroddrick) [reply src/registrar/models/domain.py:456]: I mean't that I am just doing the bare bones here, that's why its a bit extra messy. Let's go over this tomorrow for sure.

**Review by zandercymatics — APPROVED** [review]
Just a few smaller things on here (won't block for that reason), but the main functionality of this ticket is working. Nice work on all of this!
